File: bindings/java/examples/BasicExample.java
/*
 * Copyright © 2025 Dyber, Inc. All Rights Reserved.
 * QUAC 100 Post-Quantum Cryptographic Accelerator - Java SDK
 * 
 * Basic Example - Device enumeration and status
 */
package com.dyber.quac100.examples;

import com.dyber.quac100.*;
import java.util.List;

/**
 * Basic example demonstrating library initialization, device enumeration,
 * and status checking.
 */
public class BasicExample {

    public static void main(String[] args) {
        System.out.println("=== QUAC 100 Java SDK Basic Example ===\n");

        try (Library lib = Library.getInstance()) {
            // Print library version
            System.out.println("Library Version: " + Library.version());
            System.out.println("Build Info:\n" + Library.buildInfo());
            System.out.println();

            // Enumerate devices
            List<DeviceInfo> devices = lib.enumerateDevices();
            System.out.println("Found " + devices.size() + " device(s):\n");

            for (DeviceInfo info : devices) {
                System.out.println("  Device " + info.getIndex() + ":");
                System.out.println("    Model:    " + info.getModelName());
                System.out.println("    Serial:   " + info.getSerialNumber());
                System.out.println("    Firmware: " + info.getFirmwareVersion());
                System.out.println("    Slots:    " + info.getKeySlots());
                System.out.println();
            }

            if (devices.isEmpty()) {
                System.out.println("No devices found. Exiting.");
                return;
            }

            // Open first device
            Device device = lib.openFirstDevice();
            System.out.println("Opened device: " + device.getInfo().getSerialNumber());

            // Get device status
            DeviceStatus status = device.getStatus();
            System.out.println("\nDevice Status:");
            System.out.println("  Temperature:    " + status.getTemperature() + "°C");
            System.out.println("  Entropy Level:  " + status.getEntropyLevel() + "%");
            System.out.println("  Total Ops:      " + status.getTotalOperations());
            System.out.println("  Healthy:        " + status.isHealthy());
            System.out.println("  Error Count:    " + status.getErrorCount());

            // Run self-test
            System.out.println("\nRunning self-test...");
            device.selfTest();
            System.out.println("Self-test PASSED");

            // Generate some random data
            System.out.println("\nGenerating random data:");
            byte[] randomBytes = device.random().bytes(32);
            System.out.println("  32 random bytes: " + Utils.toHex(randomBytes));

            String uuid = device.random().uuid();
            System.out.println("  Random UUID:     " + uuid);

            // Close device
            device.close();
            System.out.println("\nDevice closed successfully.");

        } catch (QuacException e) {
            System.err.println("QUAC Error [" + e.getErrorCode() + "]: " + e.getMessage());
            e.printStackTrace();
        }

        System.out.println("\n=== Example Complete ===");
    }
}

File: bindings/c++/examples/basic_example.cpp
/**
 * @file basic_example.cpp
 * @brief QUAC 100 C++ SDK - Basic Example
 * @copyright Copyright © 2025 Dyber, Inc. All Rights Reserved.
 *
 * Demonstrates basic library initialization and device operations.
 */

#include <quac100/quac100.hpp>
#include <iostream>
#include <iomanip>

int main()
{
    std::cout << "QUAC 100 C++ SDK - Basic Example\n";
    std::cout << "================================\n\n";
    std::cout.flush();

    try
    {
        // Initialize library (RAII - automatically cleans up)
        std::cout << "Initializing library...\n";
        std::cout.flush();
        quac100::Library lib;

        std::cout << "Library Version: " << quac100::Library::version() << "\n";
        std::cout << "Build Info: " << quac100::Library::buildInfo() << "\n\n";
        std::cout.flush();

        // Enumerate devices
        auto devices = lib.enumerateDevices();
        std::cout << "Found " << devices.size() << " device(s)\n\n";
        std::cout.flush();

        for (const auto &info : devices)
        {
            std::cout << "Device " << info.index << ":\n";
            std::cout << "  Model: " << info.modelName << "\n";
            std::cout << "  Serial: " << info.serialNumber << "\n";
            std::cout << "  Firmware: " << info.firmwareVersion << "\n";
            std::cout << "  Key Slots: " << info.keySlots << "\n\n";
            std::cout.flush();
        }

        // Open first device
        std::cout << "Opening first device...\n";
        std::cout.flush();
        auto device = lib.openFirstDevice();
        std::cout << "Opened device successfully!\n\n";
        std::cout.flush();

        // Get device status
        auto status = device.status();
        std::cout << "Device Status:\n";
        std::cout << "  Temperature: " << status.temperature << " C\n";
        std::cout << "  Entropy Level: " << status.entropyLevel << "%\n";
        std::cout << "  Total Operations: " << status.totalOperations << "\n";
        std::cout << "  Health: " << (status.isHealthy ? "OK" : "WARNING") << "\n\n";
        std::cout.flush();

        // Run self-test
        std::cout << "Running self-test... ";
        std::cout.flush();
        device.selfTest();
        std::cout << "PASSED\n\n";
        std::cout.flush();

        // Generate some random bytes
        auto randomBytes = device.random().bytes(16);
        std::cout << "Random bytes: " << quac100::utils::toHex(randomBytes) << "\n";
        std::cout.flush();

        // Generate a UUID
        auto uuid = device.random().uuid();
        std::cout << "Random UUID: " << uuid << "\n\n";
        std::cout.flush();

        std::cout << "Basic example completed successfully!\n";
        std::cout.flush();
    }
    catch (const quac100::Exception &e)
    {
        std::cerr << "QUAC Error: " << e.what() << " (code: " << e.codeInt() << ")\n";
        std::cerr.flush();
        return 1;
    }
    catch (const std::exception &e)
    {
        std::cerr << "Standard exception: " << e.what() << "\n";
        std::cerr.flush();
        return 1;
    }

    return 0;
}